File: c++/kmp.c
#include<iostream>
#include<string>
using namespace std;

int* getNext(string str2);
int getIndexOf(string str1, string str2);

int main(){
    string str1 = "1234567fdsa";
    string str2 = "fdsa";
    int i = getIndexOf(str1, str2);
    cout << i << endl;
    return 0;
}

int getIndexOf(string str1, string str2){
    if( str1.length() < str2.length() )
        return -1;
    int* next = getNext(str2);
    int i = 0;
    int j = 0;
    while( i < str1.length() && j < str2.length() ) {
        if( str1[i] == str2[j] ) {
            i++;
            j++;
        }
        else if(next[j] == -1)
            i++;
        else
            j = next[j];
    }
    return j == str2.length() ? i - j : -1;
}

int* getNext(string str2){
    int lenOfStr2 = str2.length();
    if( lenOfStr2 == 1){
        int* next= new int[1];
        next[1] = -1;
        return next;
    }
   int* next = new int[lenOfStr2];
   next[0] = -1;
   next[1] = 0;
   int i = 2;
   int cn = 0;
   while( i < lenOfStr2 ){
       if( str2[i - 1] == str2[cn] )
           next[i++] = next[i-1] + 1;
       else{
           if( cn == -1 )
               next[i++] = 0;
           else
               cn = next[cn];
       }
   }
   return next;
}

File: python/kmp.py
def getIndexOf(str1, str2):
    if len(str1) < len(str2):
        return [-1]
    next = getNext(str2)
    i = 0
    j = 0
    while i < len(str1) and j < len(str2):
        if str1[i] == str2[j]:
            i += 1
            j += 1
        elif next[j] == -1:
            i += 1
        else:
            j = next[j]
    return i - j if j == len(str2) else -1

def getNext(str2):
    if(len(str2) == 1):
        return [-1]
    next = [0] * len(str2)
    next[0] = -1
    next[1] = 0
    i = 2
    cn = 0
    while i < len(str2):
        if str2[i-1] == str2[cn]:
            cn += 1
            next[i] = cn
            i += 1
        else:
            if cn == 0:  #跳无可跳的时候
                next[i] = 0
                i += 1
            else:
                cn = next[cn]
    return next